aten/src/ATen/native/Embedding.cpp

```diff
@@ -81,74 +69,56 @@ Tensor embedding_sparse_backward(
   return sparse_type._sparse_coo_tensor_unsafe(index, values, weight_size);
 }
 
-Tensor embedding_dense_backward_cpu(
+Tensor embedding_dense_backward(
     const Tensor & grad_, const Tensor & indices, int64_t num_weights,
     int64_t padding_idx, bool scale_grad_by_freq) {
-
+      
   auto indices_arg = TensorArg(indices, "indices", 2);
   checkScalarType("embedding_backward", indices_arg, kLong);
 
   auto indices_contig = indices.contiguous();
-  auto indices_data = indices_contig.data<int64_t>();
-  int64_t numel = indices.numel();
-
-  std::unique_ptr<int64_t[]> counts;
-  if (scale_grad_by_freq) {
-    counts.reset(new int64_t[num_weights]);
-    for (int i = 0; i < numel; i++) {
-      counts[indices_data[i]] = 0;
-    }
-    for (int i = 0; i < numel; i++) {
-      counts[indices_data[i]]++;
+  int64_t numel = indices_contig.numel();
+
+  indices_contig = indices_contig.view(-1);
+
+  Tensor counts = at::ones_like(indices_contig, grad_.type());
+  if (scale_grad_by_freq){
+    Tensor unique, inverse, mask, cnt;
+    std::tie(unique,inverse) = indices_contig._unique();
```

Comment:
Is this efficient? I think the old way was quite good except that you'd want a tensor somewhere instead of an array.

Reply:
I have posted a benchmark, please have a look.